lib/src/rewrite.rs

```diff
@@ -516,9 +503,23 @@ impl<'settings, 'repo> DescendantRebaser<'settings, 'repo> {
             .cloned()
             .collect();
 
+        let old_commits_expression =
+            RevsetExpression::commits(self.mut_repo.parent_mapping.keys().cloned().collect())
+                .union(&RevsetExpression::commits(
+                    self.mut_repo.abandoned.iter().cloned().collect(),
+                ));
+        let heads_to_add_expression = old_commits_expression
+            .parents()
+            .minus(&old_commits_expression);
+        let mut heads_to_add: HashSet<_> = heads_to_add_expression
+            .evaluate_programmatic(self.mut_repo)
+            .unwrap()
+            .iter()
+            .collect();
+
         let mut heads_to_remove = vec![];
         for old_parent_id in self.mut_repo.parent_mapping.keys() {
-            self.heads_to_add.remove(old_parent_id);
+            heads_to_add.remove(old_parent_id);
```

Comment:
nit: maybe we can remove this because `old_commits_expression` is excluded?
(might be better to do in a separate commit)

Reply:
True. I think we can make some further simplifications here too. We should also rewrite `update_references()` and `update_wc_commits()`. I'll leave all those for another PR.